src/client.rs

```diff
@@ -139,89 +87,88 @@ impl<Rt: HttpRoundTripper + 'static> Client<Rt> {
     /// Note that the requests need to have valid IDs, so it is advised to create the requests
     /// with [build_request].
     pub fn send_batch(&self, requests: &[Request]) -> Result<Vec<Option<Response>>, Error> {
-        if requests.len() < 1 {
+        if requests.is_empty() {
             return Err(Error::EmptyBatch);
         }
 
         // If the request body is invalid JSON, the response is a single response object.
         // We ignore this case since we are confident we are producing valid JSON.
-        let responses: Vec<Response> = self.send_raw(&requests)?;
+        let responses: Vec<Response> = self.transport.call(requests)
+            .map_err(|e| Error::Transport(e.into()))?;
         if responses.len() > requests.len() {
             return Err(Error::WrongBatchResponseSize);
         }
 
-        // To prevent having to clone responses, we first copy all the IDs so we can reference
-        // them easily. IDs can only be of JSON type String or Number (or Null), so cloning
-        // should be inexpensive and require no allocations as Numbers are more common.
-        let ids: Vec<serde_json::Value> = responses.iter().map(|r| r.id.clone()).collect();
+        //TODO(stevenroose) check if the server preserved order to avoid doing the mapping
```

Comment:
From [the specs](https://www.jsonrpc.org/specification#batch):
> The Server MAY process a batch rpc call as a set of concurrent tasks, processing them in any order and with any width of parallelism.
The Response objects being returned from a batch call MAY be returned in any order within the Array. The Client SHOULD match contexts between the set of Request objects and the resulting set of Response objects based on the id member within each Object.